changelogs/unreleased/gh-6329-applier-in-thread.md

```diff
@@ -0,0 +1,5 @@
+## feature/replication
+
+* Make it possible to decode incoming replication data in a separate thread. Add
+  the `replication_threads` configuration option, which controls how many
```

Comment:
> It could be good to use a more specific name for this setting. For instance, `downstream_threads`.

Maybe `replication_upstream_threads` then?
`downstream` refers to relay, while `upstream` is for appliers, and I feel like simply `upstream_threads` will scare the users. We ourselves are puzzled with `upstream` and `downstream` every time.

After a second thought, `replication_applier_threads` sounds even better, IMO. We use the `applier` name in logs, so it's fine to expose it to configuration. Is it?

> Also in the docbot request you said people can specify up to 1000 threads. Why so many? We can't have more than 32 downstreams anyway.

Actually, it's possible to have an infinite number of anonymous replicas following each other in a full-mesh (which's a quite unlikely scenario, of course).

So, it might be useful to have more than 32 threads (in some really rare cases).

I settled with 1000 just to be consistent with iproto_threads. By the way, the cfg option name was also inspired by `iproto_threads`.

> Did you make a microbenchmark to check perf difference or CPU % new distribution between threads? I tried, but my `htop` does not show threads so I can't check new CPU usage.

I made some measurements right when submitting this PR, and the numbers were roughly 200% CPU on master and ~150% cpu on replica. I didn't measure the per-thread load though.

I'll re-check that now and come back with a recipe and results.

Reply:
> Maybe `replication_upstream_threads` then?
> `downstream` refers to relay, while `upstream` is for appliers, and I feel like simply `upstream_threads` will scare the users. We ourselves are puzzled with `upstream` and `downstream` every time.

The number of relay threads isn't (and AFAIU can't possibly be) configurable so I'd go with `replication_threads`. It'd be consistent and straightforward IMO:
- `iproto_threads` - threads applying iproto requests.
- `replication_threads` - threads applying replication requests

I don't insist though.

Reply:
Wow, `htop` doesn't work on mac, indeed.
I used `ps -M` instead. It's good enough.

I ran the tests locally, and turns out I don't have enough CPU cores to run a comprehensive test with master and replica :) They both use 3 threads actively: tx, wal, and {relay/applier}, so ideally I'd load 6 cores, but I only have 4, so the load was ~70% on each of the threads.

Here's what I've done instead: run master and register replica on it.
Stop replica. Load master for some time, so it has a lot of xlogs piled up for the replica.
Stop the load and start the replica.

This way only 4 threads are busy: relay(on master), tx, applier, WAL (on replica).

Here are my results
|                  | applier fiber | applier thread     |
|------------------|---------------|--------------------|
| Relay load       | ~96 %         | ~98 %              |
| Replica sync RPS | 190k          | 195k               |
| Replica `ps -M`  | **96.7** / 0 / **51.7** / 0 | **64.0** / 0 / **65.2** / 0 / **53.2** |
| Replica total load | **148.4** | **182.4** |

I don't know which reading corresponds to which thread, unfortunately,
but it seems that the first reading is for the "tx" thread (96% vs 64 %), the last meaningful reading is for WAL (51.7% vs 53.2%) and the 3rd reading in the right column is for applier thread (65.2%).

Since master's relay is loaded to almost 100% in both tests, I assume we are limited by the relay's speed now (remember that it issues a `write` per each sent row)

Here's the test I prepared:

1. build Tarantool:
`cmake . -DCMAKE_BUILD_TYPE=Release -DENABLE_BUNDLED_LIBCURL=OFF && make -j`

2. Start master with the following script

<details><summary>test_applier_load.lua</summary>

```lua
local fiber = require('fiber')
local console = require('console')

local _running = false
local fibs = {}

local function fiber_func(idx)
    while _running do
        box.space.test:replace{idx, 'Here comes John Cena!'}
    end
end

function start()
    _running = true
    for i = 1,10 do
        local j = #fibs
        fibs[j+1] = fiber.new(fiber_func, j)
        fiber.set_joinable(fibs[j+1], true)
    end
end

function stop()
    _running = false
    for j = 1, #fibs do
        fiber.join(fibs[j])
        fibs[j] = nil
    end
    assert(#fibs == 0)
end

box.cfg{listen=3301}

box.once("bootstrap", function()
    box.schema.space.create('test')
    box.space.test:create_index('pk')
    box.schema.user.grant('guest', 'replication')
    box.snapshot()
end)

console.start()
```
</details>


3. Start replica, issue `box.cfg{replication=3301}`, stop replica.

4. Create some load on master with `start()`. Call `stop()` once the master has performed roughly 18M replaces.

5. Start replica once again, issues `box.cfg{replication=3301}` and check `ps -M` while it's syncing.